### ISSUE TITLE
Grammar

### ISSUE THREAD

**Issue by RadeKornjaca** [DaniloNovakovic/git_gud_REPL#9]
Implementing a data persistence structure for a language grammar. Grammar is defined by terms and expressions. The implementation will be contained from two parts:
- defining terms
- defining expressions
Both will be defined on a same way, only expressions will at some point use terms to define themselves.
Proposed list of terminals:
- DIGIT
- LEFT_SQUARE_BRACKET
- RIGHT_SQUARE_BRACKET
- COMMA
- DOT_PRODUCT
- CROSS_PRODUCT

Data structure that will keep grammar at one place could be dynamic array that will contain a pointer to another data structure, which could be either dynamic array or a linked list with set of terminals (which will be strings, in particular to ease down implementation, let their length be two characters long, we will revisit this at some point).

To recapitulate, this assignment will contain next steps (not necessarily in that order)
- making a linear list that contains strings (max size 2 characters)
- making a dynamic array that contains pointers to the data structures from previous step
- gluing it all up by defining an enumeration type with a proposed list of terminals which will be used instead of array indexing (enumerations values are integer values in background, so this is one of approach to make some sort of primitive table mechanism)

First and second step should result with libraries in "data_structures" repository. Third step should result with "grammar.h" and "grammar.c" file (.c file may contain very modest content, but we will expand it later).

**Comment by DaniloNovakovic** [comment]
If I get this correctly first I need to create enum{ DIGIT, LEFT_SQUARE_BRACKET, RIGHT_SQUARE_BRACKET, COMMA, DOT_PRODUCT, CROSS_PRODUCT };

Then I need to create something like this:
typedef struct node_st { char input[MAX_INPUT]; int type; struct node_st *pnext; } NODE ;
typedef struct table_st { NODE **parray, int filled, capacity; } TABLE;  //Array consting of pointers(or multiple pheads that we used in list, and then I link first node with type = DIGIT to the pointer at parray[DIGIT].  and similar. Something like hash table (We learned the idea on the theory from "Softverski alati u elektroenergetici" but I never tried implementing it. I think I can do it, gonna be fun :D )
Only question I have (idk how i didn't ask you by now in class) is how do i malloc **parray (or array of arrays) what do i need to type in sizeof( ) ? Do i need to do something like malloc(sizeof(NODE *)) in this case or what?

**Comment by RadeKornjaca** [comment]
Yes, exactly like that. Just wanted to make sure that when I say "linear list", it could be either dynamic array or linked list, so you don't have to specifically make linked lists for sets of terms, dynamic array should also do fine.

**Comment by DaniloNovakovic** [comment]
Should I have a function in grammar.c that imputs the elements/chars from stdin into our TABLE structure?Or are we gonna do that in main.c file, and if so is my solution ok then?

**Comment by RadeKornjaca** [comment]
You could make grammar elements loadable from some input file.
